README.md
Using the **YouTube scraper**, you can extract data from keyword search results, scrape detailed data on videos, like and dislike ratio and channels, download captions and scrape comment sections. 

Unlike with the [official YouTube API](https://developers.google.com/youtube/v3), with this YouTube scraper, you can scrape the results without quota limits and log in requirement. 

Our YouTube API is open-source and you can easily run it locally or on your system. Contributions are welcome.

## Features

- Scrape videos by specifying a multiple search keywords or URLs to get [video details](https://apify.com/bernardo/youtube-scraper#scraper-output), including e.g. like/dislike ratio.
- Scrape channel details (username, description, number of subscribers etc.)
- **[NEW]** Scrape and download YouTube subtitles and captions (both auto- and user-generated) in any language from any country. 
- **[NEW]** Scrape YouTube comment section (no nested comments at the moment though).

## Tutorial

For a more detailed explanation of [how to scrape YouTube](https://blog.apify.com/how-to-scrape-youtube) read a step-by-step tutorial on our [blog](https://blog.apify.com/).

And for more ideas on how to use the extracted data, check out our [industries pages](https://apify.com/industries) for concrete ways web scraping results are already being used across the projects and businesses of various scale and direction - in [media and marketing](https://apify.com/industries/marketing-and-media), for instance.

## Cost of usage

On average, scraping **1000 items** from YouTube via Apify platform will cost you around **2.5 USD credits** off your subscription plan. For more details about the plans we offer, platform credits and usage, see the [platform pricing page](https://apify.com/pricing/actors).

If you're not sure how much credits you've got left on your plan and whether you might need to upgrade, you can always check your limits in the *Settings* -> *Usage and Billing* tab in [your Console](https://console.apify.com/).  
The easiest way to know how many credits your actor will need is to perform a test run. 

### Proxy usage
This actor, as most [social media-related scrapers](https://apify.com/store?category=SOCIAL_MEDIA), requires **Proxy servers** to run properly. You can use either your own proxy servers or you can use  [Apify Proxies](https://www.apify.com/docs/proxy). We recommend using [dataset proxies](https://help.apify.com/en/articles/5265932-what-is-a-proxy) to achieve the best scraping potential of this actor.

## Input parameters

If this actor is run on our [Platform](https://console.apify.com/), a user-friendly UI there will help you out in configuring all the necessary and optional parameters of this scraper before running it. Our YouTube actor recognizes the following input fields:

 - **searchKeywords** - Your YouTube search query, say *Nimbus 2000 reviews*; this one can be used instead of a URL.
	 -  **startUrls** - A more accurate alternative to **searchKeywords**. By inserting specific URLs from YouTube you can provide search, channel or videos URLs.
 - **maxResults** - sets how many videos should be scraped from each search or channel. Defaults to 50, but you can leave it empty for unlimited search.
 - **maxComments** - Limits the number of comments that you want to scrape.  0 or empty means no comments will be scraped.

 - **downloadSubtitles** - Scrape both user-generated and auto-generated captions and convert them to SRT format. Boolean value, defaults to false.
	 - **subtitlesLanguage** - Download only subtitles of the selected language (possible values `"en"`, `"de"`, `"es"`...)
	 - **preferAutoGeneratedSubtitles** -  Prefer the autogenerated speech-to-text subtitles to the user made ones.
	 - **saveSubsToKVS** - Saves the scraped subtitles in the *Apify Key Value Store*.
 - **proxyConfiguration** *(required)* - Configures proxy settings
 - **verboseLog** *(required)* - Turns on verbose logging for accurate monitoring and having more details about the runs.
 
*See more technical details of the input parameters in the [Input Schema tab](https://apify.com/bernardo/youtube-scraper/input-schema#searchKeywords) of this actor.*

### Example

```json
{
    "searchKeywords": "Terminator dark fate",
    "maxResults": 30,
    "startUrls": [{
        "url": "https://www.youtube.com/channel/UC8w/videos" // channel videos
    }, {
        "url": "https://www.youtube.com/results?search_query=finances" // search queries
    }, {
        "url": "https://www.youtube.com/watch?v=kJQP7kiw5Fk" // videos
    }],
    "proxyConfiguration": {
        "useApifyProxy": true
    },
    "verboseLog": false
}

```


## YouTube Scraper output

After the actor finishes the run, it will store the scraped results in a the *Dataset*. Each YouTube video becomes a separate record in the dataset  (see a JSON example below). Using the Apify platform, you can choose to present and download the contents of the dataset in different data formats (JSON, RSS, XML, HTML Table...).

### Example
```json
{
  "title": "Terminator: Dark Fate - Official Trailer (2019) - Paramount Pictures",
  "id": "oxy8udgWRmo",
  "url": "https://www.youtube.com/watch?v=oxy8udgWRmo",
  "viewCount": 15432,
  "date": "2019-08-29T00:00:00+00:00",
  "likes": 121000,
  "dislikes": 23000,
  "channelName": "Paramount Pictures",
  "channelUrl": "https://www.youtube.com/channel/UCF9imwPMSGz4Vq1NiTWCC7g",
  "numberOfSubscribers": 1660000,
  "details": "Welcome to the day after <a class=\"yt-simple-endpoint style-sco..."
}

```
See the  [Apify API reference](https://www.apify.com/docs/api)  to learn in more detail about getting results from this YouTube Scraper.

### **How can you use the data extracted from YouTube:**

- **Compile reviews of products and services** - make purchasing and investment decisions backed by data.

- **Monitor YouTube for brand awareness** - keep track of brand mentions, audience reach and web reputation.

- **Estimate the impact of YouTube campaigns** - estimate ROI for advertisement or referrals from YouTube channels and scale marketing campaigns accordingly.

-  **Apply scraped data in journalism** - track down and tackle fake news, bot activity, as well as illegal, misleading or harmful content. Dissect big news topics and analyze sentiment on web.
-   **Collect data for any kind of research** -  identify and follow emerging trends or topics and even predict the new ones: globally or by country and language.
## Changelog

You can see all newest changes to this YouTube scraper listed in this  [CHANGELOG.md](https://github.com/bernardro/actor-youtube-scraper/blob/master/CHANGELOG.md) file.

## Notes for the developers on customizing the actor
Here are the calculations for a typical resource usage of YouTube Scraper on Apify platform:

| **Resource** | **Average** | **Max** |
|--|--|--|
|  Memory| 480.3 MB | 1.1 GB |
| CPU | 53% | 140% |

This actor uses xPaths to find DOM elements; they are all stored in one file for easy update. All xPath variables and functions end in 'Xp'.

### Extend output function

Extend output function allows you to omit output, add some extra properties to the output by using the  `page`  variable or change the shape of your output altogether:

```js
async ({ item }) => {
    // remove information from the item
    item.details = undefined;
    // or delete item.details;
    return item;
}
```

```js
async ({ item, page }) => {
    // add more info, in this case, the shortLink for the video
    const shortLink = await page.evaluate(() => {
        const link = document.querySelector('link[rel="shortlinkUrl"]');
        if (link) {
            return link.href;
        }
    });

    return {
        ...item,
        shortLink,
    }
}
```

```js
async ({ item }) => {
    // omit item, just return null
    return null;
}
```

### Extend scraper function

Extend scraper function allows you to add functionality to the existing baseline behavior. For example, you may enqueue related videos, but not recursively:

```js
async ({ page, request, requestQueue, customData, Apify }) => {
    if (request.userData.label === 'DETAIL' && !request.userData.isRelated) {
        await page.waitForSelector('ytd-watch-next-secondary-results-renderer');

        const related = await page.evaluate(() => {
            return [...document.querySelectorAll('ytd-watch-next-secondary-results-renderer a[href*="watch?v="]')].map(a => a.href);
        });

        for (const url of related) {
            await requestQueue.addRequest({
                url,
                userData: {
                    label: 'DETAIL',
                    isRelated: true,
                },
            });
        }
    }
}
```

*NB: If this specific function throws an exception, it will retry the same URL it was visiting again.*

## Acknowledgments and personal data

This scraper collects cookies and privacy consent dialogs on your behalf. Therefore, you should be aware that the results from your YouTube scraping might contain personal data. 

Personal data is protected by GDPR ([EU Regulation 2016/679](https://eur-lex.europa.eu/eli/reg/2016/679/oj)),  and by other regulations around the world. You should **not** scrape personal data unless you have a legitimate reason to do so. 

If you're unsure whether your reason is legitimate, consult your lawyers. You can also read our blog post on the  [legality of web scraping](https://blog.apify.com/is-web-scraping-legal/).

## Other video and social media scrapers

We have other video-related scrapers in stock for you; to see more of those, check out the [Video Category in Apify Store](https://apify.com/store?category=VIDEOS) or the compilation of [Social Media Scrapers](https://apify.com/store?category=SOCIAL_MEDIA). 

## Your feedback

We’re always working on improving the performance of our actors. So if you’ve got any technical feedback about the work of our YouTube API, or simply **found a bug,** please create an issue on the [Github page](https://github.com/bernardro/actor-youtube-scraper) and we’ll get to it.



CHANGELOG.md
## 2022-07-20
*Fixes*
- Correctly handle videos with comments turned off. 
- Add `commentsTurnedOff` to output.

## 2022-06-10
*Fixes*:
- Channel page without `/watch` selector

## 2021-09-15
*Features*
- Add possibility to scrape video comments. See `maxComments` input field.

2021-06-16
*Features*
- Revamped subtitles downloading - added possibility to download all available subtitles
  (availability defined by languages) and to prefer automatically generated subtitles before the user generated
  ones.


2021-06-14
*Features*:
- Add subtitle type to output (extendedOutputFunction). **Note**: You must set `downloadSubtitles` variable to `true` for this
  feature to take effect.

2021-06-11
*Features*:
- Subtitles are now downloadable (saved to KeyValueStore as `videoID_languageCode`)

2021-05-21
*Features*:
- Update SDK

*Fixes*
- Random zero results when searching
- Click consent dialog

2021-04-14
*Fixes*
- Fixed changed selector that completely prevented the scrape

2021-03-21
*Features*:
- Updated SDK version for session pool changes
- Add `handlePageTimeoutSecs` parameter to INPUT_SCHEMA


2021-03-15
*Fixes:*
- Fixed selector causing no data scraped
- Removed stealth causing issues with new layout

2020-09-27
- Increased waiting timeouts to better handle concurrency
- Added saving screenshots on errors
- Better handling of Captchas, a page is automatically retried and the browser is restarted with a new proxy
- `verboseLog` is off by default
- Added info how many videos were enqueued and overall better logging


test/helpers/chai.js
const chai = require('chai');

chai.config.includeStack = true;

global.expect = chai.expect;
global.AssertionError = chai.AssertionError;
global.Assertion = chai.Assertion;
global.assert = chai.assert;


src/utility.js
const moment = require('moment');
const Apify = require('apify');
const Puppeteer = require('puppeteer'); // eslint-disable-line
const vm = require('vm');

const { log, sleep } = Apify.utils;

const CONSTS = require('./consts');

exports.handleErrorAndScreenshot = async (page, e, errorName) => {
    await Apify.utils.puppeteer.saveSnapshot(page, { key: `ERROR-${errorName}-${Math.random()}` });
    throw `Error: ${errorName} - Raw error: ${e.message}`;
};

/**
 * @param {{
 *  requestQueue: Apify.RequestQueue,
 *  page: Puppeteer.Page,
 *  maxRequested: number,
 *  isSearchResultPage: boolean,
 *  searchOrUrl: string,
 * }} loadVideosUrlsParams
 */

exports.loadVideosUrls = async (loadVideosUrlsParams) => {
    const { requestQueue, page, maxRequested, isSearchResultPage, searchOrUrl } = loadVideosUrlsParams;
    const { youtubeVideosSection, youtubeVideosRenderer, url } = CONSTS.SELECTORS.SEARCH;

    log.debug('loadVideosUrls', { maxRequested });
    let shouldContinue = true;
    let videosEnqueued = 0;
    let videosEnqueuedUnique = 0;

    const logInterval = setInterval(
        () => log.info(`[${searchOrUrl}]: Scrolling state - Enqueued ${videosEnqueuedUnique} unique video URLs, ${videosEnqueued} total`),
        60000,
    );

    try {
        while (shouldContinue) { // eslint-disable-line no-constant-condition
            // youtube keep adding video sections to the page on scroll
            await page.waitForSelector(youtubeVideosSection);
            const videoSections = await page.$$(youtubeVideosSection);

            log.debug('Video sections', { shouldContinue, videoSections: videoSections.length });
            let videoCount = 0;

            for (const videoSection of videoSections) {
                // each section have around 20 videos
                await page.waitForSelector(youtubeVideosRenderer);
                const videos = await videoSection.$$(youtubeVideosRenderer);

                log.debug('Videos count', { shouldContinue, videos: videos.length });

                for (const video of videos) {
                    try {
                        await video.hover();
                    } catch (e) {
                        log.debug('Video hover error', { e: e.message });
                    }

                    // sometimes a single $eval will make this call crash, and will stop the loop
                    // keeping it as $$eval always return an array, and an empty one when it's not found
                    const videoUrls = await video.$$eval(url, (els) => els.map((el) => el.href));

                    if (!videoUrls.length) {
                        log.debug('Video url not found');
                        continue; // eslint-disable-line no-continue
                    }

                    const rq = await requestQueue.addRequest({
                        url: videoUrls[0],
                        userData: { label: 'DETAIL' },
                    });

                    videosEnqueued++;

                    if (!rq.wasAlreadyPresent) {
                        // count only unique videos
                        videosEnqueuedUnique++;
                    }

                    if (videosEnqueued >= maxRequested) {
                        shouldContinue = false;
                        break;
                    }

                    await sleep(CONSTS.DELAY.HUMAN_PAUSE.max);

                    if (!isSearchResultPage) {
                        // remove the link on channels, so the scroll happens
                        await video.evaluate((el) => el.remove());
                    }

                    videoCount++;
                }

                await sleep(CONSTS.DELAY.START_LOADING_MORE_VIDEOS);

                if (isSearchResultPage) {
                    // remove element after extracting result urls. removing it make the page scroll,
                    // and frees up memory. only delete nodes in search results
                    await videoSection.evaluate((el) => el.remove());
                }

                if (!shouldContinue) {
                    break;
                }
            }

            if (!videoCount) {
                shouldContinue = false;
                break;
            }
        }
    } catch (e) {
        clearInterval(logInterval);
        throw e;
    }
    clearInterval(logInterval);
    log.info(`[${searchOrUrl}]: Scrolling finished - Enqueued ${videosEnqueuedUnique} unique video URLs, ${videosEnqueued} total`);
};

exports.getDataFromXpath = async (page, xPath, attrib) => {
    await page.waitForXPath(xPath, { timeout: 120000 });
    const xElement = await page.$x(xPath);
    return page.evaluate((el, key) => el[key], xElement[0], attrib);
};

exports.getDataFromSelector = async (page, slctr, attrib) => {
    const slctrElem = await page.waitForSelector(slctr, { visible: true, timeout: 60000 });
    return page.evaluate((el, key) => el[key], slctrElem, attrib);
};

/**
 * @param {string} url
 */
exports.categorizeUrl = (url) => {
    try {
        const pUrl = new URL(url, 'https://www.youtube.com');

        if (!pUrl.hostname.includes('youtube.com')) {
            throw new Error('Invalid youtube url');
        }

        let label = 'MASTER';

        if (pUrl.searchParams.get('v')) {
            label = 'DETAIL';
        } else if (pUrl.searchParams.get('search_query')) {
            label = 'SEARCH';
        } else if (pUrl.pathname.includes('/channel/') || pUrl.pathname.includes('/user/') || pUrl.pathname.includes('/c/')) {
            label = 'CHANNEL';
        }

        return label;
    } catch (e) {
        log.exception(e, 'categorizeUrl', { url });
        return null;
    }
};

exports.unformatNumbers = (numStr) => {
    const numberMatch = numStr.replace(/[^0-9,.]/ig, '');
    if (numberMatch) {
        const number = parseFloat(numberMatch.replace(/,/g, ''));
        const multiplierMatch = numStr.match(/(?<=[0-9 ])[mkb]/ig);

        if (multiplierMatch) {
            const multiplier = multiplierMatch[0].toUpperCase();
            switch (multiplier) {
                case 'K': {
                    return Math.round(number * 1000);
                }
                case 'M': {
                    return Math.round(number * 1000000);
                }
                case 'B': {
                    return Math.round(number * 1000000000);
                }
                default: throw new Error('Unhandled multiplier in getExpandedNumbers');
            }
        }

        return number;
    }

    // some videos may not have likes, views or channel subscribers
    return 0;
};

exports.moveMouseToElemXp = async (pptPage, xPath, mouseMoveSteps, name) => {
    const targetElem = await pptPage.waitForXPath(xPath, { visible: true });
    if (targetElem.length > 0) {
        log.debug(`${name} found at ${xPath}`);
    }

    const searchBoxRect = await targetElem.boundingBox();
    const { xPos, yPos } = exports.getRandClickPos(searchBoxRect);

    // pause like real user
    await Apify.utils.sleep(exports.getDelayMs(CONSTS.DELAY.HUMAN_PAUSE));

    // move mouse to target
    await pptPage.mouse.move(xPos, yPos, { steps: mouseMoveSteps });
};

exports.moveMouseToCenterScreen = async (pptPage, mouseMoveSteps) => {
    const viewPort = await pptPage.viewport();
    const { width, height } = viewPort;

    const xPos = Math.ceil(width / 2);
    const yPos = Math.ceil(height / 2);

    await pptPage.mouse.move(xPos, yPos, { steps: mouseMoveSteps });
};

exports.clickHoveredElem = async (pptPage, xPath) => {
    log.debug(`clicking on ${xPath}`);
    await pptPage.mouse.down();
    await Apify.utils.sleep(exports.getDelayMs(CONSTS.DELAY.MOUSE_CLICK));
    await pptPage.mouse.up();
};

/**
 * @param {Puppeteer.Page} pptPage
 * @param {string[]} keywords
 */
exports.doTextInput = async (pptPage, keywords) => {
    for (let i = 0; i < keywords.length; i++) {
        await pptPage.type('input#search', keywords[i], { delay: CONSTS.DELAY.BTWN_KEY_PRESS.max });
        await Apify.utils.sleep(exports.getDelayMs(CONSTS.DELAY.BTWN_KEY_PRESS));
    }
};

exports.getCutoffDate = (historyString) => {
    // input should have been validated
    // in format 'x minutes/hours/days/weeks/months/year ago'
    const matchNumbers = historyString.match(new RegExp('[0-9]+', 'ig'));
    const numDurations = parseInt(matchNumbers[0], 10);

    const matchDuration = historyString.match(new RegExp('(minute|hour|day|week|month|year)', 'ig'));
    const durationStr = matchDuration[0].toLowerCase();
    const durationType = `${durationStr}s`;

    return moment().subtract(numDurations, durationType);
};

/**
 * @param {string} postsFromDate
 */
exports.isDateInputValid = (postsFromDate) => {
    if (postsFromDate) {
        const matches = postsFromDate.match(/(^(1|([^0a-zA-Z ][0-9]{0,3})) (minute|hour|day|week|month|year))s?/ig);
        return !!matches;
    }

    return false;
};

/**
 * @param {string} postsFromDate
 */
exports.getYoutubeDateFilters = (postsFromDate) => {
    if (!exports.isDateInputValid(postsFromDate)) {
        return [];
    }

    const now = moment();
    const then = exports.getCutoffDate(postsFromDate);
    const duration = moment.duration(now.diff(then));

    const dateFilterMap = [
        { inTheLast: 'years', filter: null },
        { inTheLast: 'months', filter: 'This year' },
        { inTheLast: 'weeks', filter: 'This month' },
        { inTheLast: 'days', filter: 'This week' },
        { inTheLast: 'hours', filter: 'Today' },
        { inTheLast: 'minutes', filter: 'Last hour' },
    ];

    const youtubeFilters = [];
    // start with the longest duration
    for (let i = 0; i < dateFilterMap.length; i++) {
        const durType = dateFilterMap[i].inTheLast;
        const durCount = duration.as(durType);

        // e.g '2 days ago' results in 'This week' youtube filter being set
        if (durCount > 1) {
            if (dateFilterMap[i].filter) {
                youtubeFilters.push(dateFilterMap[i].filter);
            }
            break;
        }
    }

    if (youtubeFilters.length > 0) {
        // if we are using any of the youtube date filters
        // then we must also sort results by 'Upload date'
        youtubeFilters.unshift('Upload date');
    }

    return youtubeFilters;
};

exports.getVideoId = (videoUrl) => {
    const matches = videoUrl.match(/(?<=[/=])([0-9A-Za-z_-]{10,})[#/]?/ig);
    if (matches && matches.length > 0) {
        for (let i = 0; i < matches.length; i++) {
            const match = matches[i];

            // two-stage regex to cover all current types of youtube URLs
            const matches2 = match.match(/[#/]/ig);
            if (!matches2 && match.length <= 12) {
                return match;
            }
        }
    }

    return '';
};

exports.getRandClickPos = (rect) => {
    const xBuffer = 0.7;
    const yBuffer = 0.4;

    // define an area well within the element's borders
    // we will be mouse-clicking within this region
    const clickableWidth = Math.ceil(xBuffer * (rect.width));
    const clickableHeight = Math.ceil(yBuffer * (rect.height));

    const randXoffset = exports.getRandBetween(1, clickableWidth);
    const randYoffset = exports.getRandBetween(1, clickableHeight);

    return { xPos: (rect.x + randXoffset), yPos: (rect.y + randYoffset) };
};

exports.isErrorStatusCode = (statusCode) => {
    if (statusCode) {
        const statusNum = parseInt(`${statusCode}`, 10);
        return statusNum >= 400;
    }

    return false;
};

exports.getMaxVideos = (numOfVideos, userMaximum) => {
    if (userMaximum > 0) {
        if (userMaximum > numOfVideos) {
            log.info(`user requested ${userMaximum} videos but only ${numOfVideos} were loaded`);
            return numOfVideos;
        }

        return userMaximum;
    }

    // no user imposed limit
    return numOfVideos;
};

exports.getRandBetween = (min, max) => {
    return Math.floor(Math.random() * (max - min + 1) + min);
};

exports.getDelayMs = (minMax) => {
    return exports.getRandBetween(minMax.MIN, minMax.MAX);
};

/**
 * @template T
 * @typedef {T & { Apify: Apify, customData: any, request: Apify.Request }} PARAMS
 */

/**
 * Compile a IO function for mapping, filtering and outputing items.
 * Can be used as a no-op for interaction-only (void) functions on `output`.
 * Data can be mapped and filtered twice.
 *
 * Provided base map and filter functions is for preparing the object for the
 * actual extend function, it will receive both objects, `data` as the "raw" one
 * and "item" as the processed one.
 *
 * Always return a passthrough function if no outputFunction provided on the
 * selected key.
 *
 * @template RAW
 * @template {{ [key: string]: any }} INPUT
 * @template MAPPED
 * @template {{ [key: string]: any }} HELPERS
 * @param {{
 *  key: string,
 *  map?: (data: RAW, params: PARAMS<HELPERS>) => Promise<MAPPED>,
 *  output?: (data: MAPPED, params: PARAMS<HELPERS> & { data: RAW, item: MAPPED }) => Promise<void>,
 *  filter?: (obj: { data: RAW, item: MAPPED }, params: PARAMS<HELPERS>) => Promise<boolean>,
 *  input: INPUT,
 *  helpers: HELPERS,
 * }} params
 * @return {Promise<(data: RAW, args?: Record<string, any>) => Promise<void>>}
 */
const extendFunction = async ({
    key,
    output,
    filter,
    map,
    input,
    helpers,
}) => {
    /**
     * @type {PARAMS<HELPERS>}
     */
    const base = {
        ...helpers,
        Apify,
        customData: input.customData || {},
    };

    const evaledFn = (() => {
        // need to keep the same signature for no-op
        if (typeof input[key] !== 'string' || input[key].trim() === '') {
            return new vm.Script('({ item }) => item');
        }

        try {
            return new vm.Script(input[key], {
                lineOffset: 0,
                produceCachedData: false,
                displayErrors: true,
                filename: `${key}.js`,
            });
        } catch (e) {
            throw new Error(`"${key}" parameter must be a function`);
        }
    })();

    /**
     * Returning arrays from wrapper function split them accordingly.
     * Normalize to an array output, even for 1 item.
     *
     * @param {any} value
     * @param {any} [args]
     */
    const splitMap = async (value, args) => {
        const mapped = map ? await map(value, args) : value;

        if (!Array.isArray(mapped)) {
            return [mapped];
        }

        return mapped;
    };

    return async (data, args) => {
        const merged = { ...base, ...args };

        for (const item of await splitMap(data, merged)) {
            if (filter && !(await filter({ data, item }, merged))) {
                continue; // eslint-disable-line no-continue
            }

            const result = await (evaledFn.runInThisContext()({
                ...merged,
                data,
                item,
            }));

            for (const out of (Array.isArray(result) ? result : [result])) {
                if (output) {
                    if (out !== null) {
                        await output(out, { ...merged, data, item });
                    }
                    // skip output
                }
            }
        }
    };
};

exports.extendFunction = extendFunction;

/**
 * Do a generic check when using Apify Proxy
 *
 * @typedef params
 * @property {any} [params.proxyConfig] Provided apify proxy configuration
 * @property {boolean} [params.required] Make the proxy usage required when running on the platform
 * @property {string[]} [params.blacklist] Blacklist of proxy groups, by default it's ['GOOGLE_SERP']
 * @property {boolean} [params.force] By default, it only do the checks on the platform. Force checking regardless where it's running
 * @property {string[]} [params.hint] Hint specific proxy groups that should be used, like SHADER or RESIDENTIAL
 *
 * @param {params} params
 * @returns {Promise<Apify.ProxyConfiguration | undefined>}
 */
module.exports.proxyConfiguration = async ({
    proxyConfig,
    required = true,
    force = Apify.isAtHome(),
    blacklist = ['GOOGLESERP'],
    hint = [],
}) => {
    const configuration = await Apify.createProxyConfiguration(proxyConfig);

    // this works for custom proxyUrls
    if (Apify.isAtHome() && required) {
        if (!configuration
            || (!configuration.usesApifyProxy
                && (!configuration.proxyUrls || !configuration.proxyUrls.length)) || !configuration.newUrl()) {
            throw '\n=======\nWrong Input! You must use Apify proxy or custom proxies with this scraper!\n\n=======';
        }
    }

    // check when running on the platform by default
    if (force) {
        // only when actually using Apify proxy it needs to be checked for the groups
        if (configuration && configuration.usesApifyProxy) {
            if (blacklist.some((blacklisted) => (configuration.groups || []).includes(blacklisted))) {
                throw '\n=======\nThese proxy groups cannot be used in this actor.'
                    + `Choose other group or contact support@apify.com to give you proxy trial:\n\n*  ${blacklist.join('\n*  ')}\n\n=======`;
            }

            // specific non-automatic proxy groups like RESIDENTIAL, not an error, just a hint
            if (hint.length && !hint.some((group) => (configuration.groups || []).includes(group))) {
                Apify.utils.log.info(`\n=======\nYou can pick specific proxy groups for better experience:\n\n*  ${hint.join('\n*  ')}\n\n=======`);
            }
        }
    }

    return configuration;
};
/**
 * Scrape video comments from video detail page.
 * @param {Puppeteer.Page} page
 * @param {number} maxComments: Maximum number of comments to scrape.
 * @returns {Promise<*>}
 */
module.exports.getVideoComments = async (page, maxComments) => {
    // This is copied from SDK - We needed to add stopScrollCallback function parameter to quit scrolling when we have
    // enough comments scraped. This should be replaced by Apify.utils.puppeteer.infiniteScroll when the SDK will be
    // updated by stopScrollCallback feature.
    const infiniteScroll = async (page, options = {}) => {
        const { timeoutSecs = 0, waitForSecs = 4, scrollDownAndUp = false, buttonSelector, stopScrollCallback } = options;
        let finished;
        const startTime = Date.now();
        const CHECK_INTERVAL_MILLIS = 1000;
        const SCROLL_HEIGHT_IF_ZERO = 10000;
        const maybeResourceTypesInfiniteScroll = ['xhr', 'fetch', 'websocket', 'other'];
        const resourcesStats = {
            newRequested: 0,
            oldRequested: 0,
            matchNumber: 0,
        };
        page.on('request', (msg) => {
            if (maybeResourceTypesInfiniteScroll.includes(msg.resourceType())) {
                resourcesStats.newRequested++;
            }
        });
        const checkFinished = setInterval(() => {
            if (resourcesStats.oldRequested === resourcesStats.newRequested) {
                resourcesStats.matchNumber++;
                if (resourcesStats.matchNumber >= waitForSecs) {
                    clearInterval(checkFinished);
                    finished = true;
                    return;
                }
            } else {
                resourcesStats.matchNumber = 0;
                resourcesStats.oldRequested = resourcesStats.newRequested;
            }
            // check if timeout has been reached
            if (timeoutSecs !== 0 && (Date.now() - startTime) / 1000 > timeoutSecs) {
                clearInterval(checkFinished);
                finished = true;
            }
        }, CHECK_INTERVAL_MILLIS);
        const doScroll = async () => {
            /* istanbul ignore next */
            await page.evaluate(async (scrollHeightIfZero) => {
                const delta = document.body.scrollHeight === 0 ? scrollHeightIfZero : document.body.scrollHeight;
                window.scrollBy(0, delta);
            }, SCROLL_HEIGHT_IF_ZERO);
        };
        const maybeClickButton = async () => {
            const button = await page.$(buttonSelector);
            // Box model returns null if the button is not visible
            if (button && await button.boxModel()) {
                await button.click({ delay: 10 });
            }
        };
        while (!finished) {
            await doScroll();
            await page.waitForTimeout(50);
            if (scrollDownAndUp) {
                await page.evaluate(() => {
                    window.scrollBy(0, -1000);
                });
            }
            if (buttonSelector) {
                await maybeClickButton();
            }
            if (stopScrollCallback) {
                if (await stopScrollCallback()) {
                    break;
                }
            }
        }
    };

    const commentSelector = 'ytd-comment-thread-renderer';
    // Scroll first to load at lease one comment.
    await page.evaluate(() => {
        window.scrollBy(0, 500);
    });
    await page.waitForSelector('ytd-comment-thread-renderer');
    await infiniteScroll(page, { stopScrollCallback: async () => {
        const commentCount = await page.evaluate((commentSelector) => {
            return document.body.querySelectorAll(commentSelector).length;
        }, commentSelector);
        log.debug(`Got ${commentCount}/${maxComments} comments for ${page.url()}`);
        return commentCount >= maxComments;
    } });
    const comments = await page.evaluate((max, commentSelector) => {
        const elements = document.body.querySelectorAll(commentSelector);
        const a = [];
        for (let i = 0; i < elements.length; i++) {
            const e = elements[i];
            const author = e.querySelector('#author-text > span').innerHTML.trim()
                .replace(/\\n/g, '');
            if (author) {
                const text = e.querySelector('#content-text').innerHTML.trim()
                    .replace(/\\n/g, '');
                a.push({
                    author,
                    comment: text,
                });
            }
            if (a.length >= max) {
                break;
            }
        }
        return a;
    }, maxComments, commentSelector);

    log.info(`Scraped ${comments.length} comments for video ${page.url()}`);

    return comments;
};
